File: C_Frog_Freddy.py
def find_final_leaf(N, M, leaves, commands):
    # Step 1: Sort leaves for efficient search
    leaves.sort()  # Sort leaves by x, then by y if x is the same
    
    # Step 2: Initialize Freddy's starting position
    current_position = leaves[0]
    leaf_set = set(leaves)
    leaf_set.remove(current_position)
    
    # Step 3: Process each command
    for command in commands:
        a, b = current_position
        new_position = None
        
        if command == 'A':
            # Move to (a + Z, b + Z) where Z > 0
            for Z in range(1, 100001):
                new_position = (a + Z, b + Z)
                if new_position in leaf_set:
                    current_position = new_position
                    leaf_set.remove(current_position)  # Leaf sinks
                    break
        elif command == 'B':
            # Move to (a + Z, b - Z) where Z > 0
            for Z in range(1, 100001):
                new_position = (a + Z, b - Z)
                if new_position in leaf_set:
                    current_position = new_position
                    leaf_set.remove(current_position)
                    break
        elif command == 'C':
            # Move to (a - Z, b + Z) where Z > 0
            for Z in range(1, 100001):
                new_position = (a - Z, b + Z)
                if new_position in leaf_set:
                    current_position = new_position
                    leaf_set.remove(current_position)
                    break
        elif command == 'D':
            # Move to (a - Z, b - Z) where Z > 0
            for Z in range(1, 100001):
                new_position = (a - Z, b - Z)
                if new_position in leaf_set:
                    current_position = new_position
                    leaf_set.remove(current_position)
                    break
        
    # Step 4: Return Freddy's final position
    return current_position

# Read inputs from the user
N, M = map(int, input())
commands = input()
leaves = []


# Read the positions of the leaves
for _ in range(N):
    x, y = map(int, input().split())
    leaves.append((x, y))

# Call the function to find the final leaf position
final_position = find_final_leaf(N, M, leaves, commands)

# Output the final position
print(final_position)

File: C_Frog_Freddy.cpp
#include <iostream>
#include <unordered_set>
#include <string>
#include <vector>
#include <tuple>

using namespace std;

struct Leaf {
    int x, y;
    bool operator==(const Leaf& other) const {
        return x == other.x && y == other.y;
    }

    struct Hash {
        size_t operator()(const Leaf& leaf) const {
            return hash<int>()(leaf.x) ^ hash<int>()(leaf.y);
        }
    };
};

int main() {
    int N, M;
    cin >> N >> M;
    
    string commands;
    cin >> commands;

    unordered_set<Leaf, Leaf::Hash> leafSet;

    vector<Leaf> leaves(N);
    for (int i = 0; i < N; i++) {
        cin >> leaves[i].x >> leaves[i].y;
        leafSet.insert(leaves[i]);
    }

    Leaf current_position = leaves[0];
    leafSet.erase(current_position); 

    for (char command : commands) {
        int a = current_position.x;
        int b = current_position.y;

        bool moved = false;

        switch (command) {
            case 'A': 
                for (int Z = 1; Z <= 100000 && !moved; Z++) {
                    Leaf nextLeaf = {a + Z, b + Z};
                    if (leafSet.count(nextLeaf)) {
                        current_position = nextLeaf;
                        leafSet.erase(current_position);  
                        moved = true;
                    }
                }
                break;

            case 'B': 
                for (int Z = 1; Z <= 100000 && !moved; Z++) {
                    Leaf nextLeaf = {a + Z, b - Z};
                    if (leafSet.count(nextLeaf)) {
                        current_position = nextLeaf;
                        leafSet.erase(current_position);
                        moved = true;
                    }
                }
                break;

            case 'C': 
                for (int Z = 1; Z <= 100000 && !moved; Z++) {
                    Leaf nextLeaf = {a - Z, b + Z};
                    if (leafSet.count(nextLeaf)) {
                        current_position = nextLeaf;
                        leafSet.erase(current_position);
                        moved = true;
                    }
                }
                break;

            case 'D': 
                for (int Z = 1; Z <= 100000 && !moved; Z++) {
                    Leaf nextLeaf = {a - Z, b - Z};
                    if (leafSet.count(nextLeaf)) {
                        current_position = nextLeaf;
                        leafSet.erase(current_position);
                        moved = true;
                    }
                }
                break;

            default:
                continue;
        }
    }

    cout << current_position.x << " " << current_position.y << endl;

    return 0;
}
